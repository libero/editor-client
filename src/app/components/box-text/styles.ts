import { makeStyles } from '@material-ui/core/styles';

export const useBoxTextEditorStyles = makeStyles((theme) => ({
  boxContainer: {
    display: 'flex',
    alignItems: 'flex-start'
  },
  editButton: {
    flexGrow: 0,
    padding: 6,
    marginLeft: theme.spacing(2),
    marginTop: theme.spacing(5)
  }
}));
