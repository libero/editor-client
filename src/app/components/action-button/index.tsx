import React from 'react';
import { Button, PropTypes } from '@material-ui/core';
import { mapValues } from 'lodash';
import AddIcon from '@material-ui/icons/Add';

import { useActionButtonStyles } from './styles';

type MuiButtonVariant = 'text' | 'outlined' | 'contained';

const VARIANTS = {
  containedWarning: {
    variant: 'contained' as MuiButtonVariant,
    classes: {
      root: 'containedRedButton'
    }
  },
  outlinedWarning: {
    variant: 'outlined' as MuiButtonVariant,
    classes: {
      root: 'outlinedRedButton'
    }
  },
  primaryContained: {
    variant: 'contained' as MuiButtonVariant,
    color: 'primary',
    classes: {}
  },
  secondaryOutlined: {
    variant: 'outlined',
    classes: {}
  },
  addEntity: {
    size: 'small',
    classes: { root: 'addEntityButton_root', startIcon: 'addEntityButton_icon' },
    variant: 'text',
    startIcon: <AddIcon />
  }
} as {
  [key: string]: {
    variant: MuiButtonVariant;
    classes: Record<string, string>;
    color?: PropTypes.Color;
  };
};

export type VariantType = keyof typeof VARIANTS;

interface ActionButtonProps {
  onClick: () => void;
  title: string;
  variant: VariantType;
  className?: string;
}

export const ActionButton: React.FC<ActionButtonProps> = (props) => {
  const classes = useActionButtonStyles();
  const variantProps = VARIANTS[props.variant];
  const variantClasses = mapValues(variantProps.classes, (className: string) => classes[className]);
  const buttonProps = { ...variantProps, classes: variantClasses };
  return (
    <Button onClick={props.onClick} {...buttonProps} className={props.className}>
      {props.title}
    </Button>
  );
};
