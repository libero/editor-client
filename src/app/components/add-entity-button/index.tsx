import React from 'react';
import AddIcon from '@material-ui/icons/Add';
import { Button } from '@material-ui/core';
import { useAddEntityBtnStyles } from './styles';

export const AddEntityButton: React.FC = () => {
  const classes = useAddEntityBtnStyles();
  return (
    <Button
      size="small"
      classes={{ root: classes.root, startIcon: classes.icon }}
      variant="text"
      startIcon={<AddIcon />}
    >
      Add Author
    </Button>
  );
};
