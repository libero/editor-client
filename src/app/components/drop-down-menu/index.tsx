import React, { useCallback, useRef } from 'react';
import { Popper, ClickAwayListener, Paper, MenuList, Button, MenuItem } from '@material-ui/core';
import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown';

import { useDropDownStyles } from './styles';

export interface DropDownMenuItemProps {
  title: string;
  enabled: boolean;
  selected?: boolean;
  action(): void;
}

export interface DropDownMenuProps {
  title: string;
  entries: DropDownMenuItemProps[];
}

export const DropDownMenu: React.FC<DropDownMenuProps> = ({ title, entries }) => {
  const [open, setOpen] = React.useState(false);
  const menuRef = useRef();
  const handleOpenMenuClick = useCallback(() => setOpen(true), [setOpen]);
  const handleMenuClose = useCallback(() => setOpen(false), [setOpen]);
  const classes = useDropDownStyles();

  const handleMenuItemClick = useCallback((action) => {
    if (action) {
      action();
    }
    setOpen(false);
  }, []);

  const menuItems = entries.map((entry, index) => {
    return (
      <MenuItem
        disabled={!entry.enabled}
        key={index}
        selected={entry.selected}
        onClick={handleMenuItemClick.bind(null, entry.action)}
      >
        {entry.title}
      </MenuItem>
    );
  });

  return (
    <div>
      <Button ref={menuRef} aria-haspopup="true" onClick={handleOpenMenuClick} className={classes.button}>
        {title}
        <ArrowDropDownIcon />
      </Button>
      <Popper anchorEl={menuRef.current} role={undefined} transition disablePortal open={open}>
        <Paper>
          <ClickAwayListener onClickAway={handleMenuClose}>
            <MenuList autoFocusItem={open}>{menuItems}</MenuList>
          </ClickAwayListener>
        </Paper>
      </Popper>
    </div>
  );
};
