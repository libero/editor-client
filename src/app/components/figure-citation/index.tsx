import React from 'react';
import ReactDOM from 'react-dom';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import { EditorView, NodeView } from 'prosemirror-view';
import { ThemeProvider } from '@material-ui/core/styles';

import { theme } from '../../styles/theme';
import { FigureCitationEditorPopup, FiguresListEntry, UNLABELLED_FIGURE_TEXT } from './figure-citation-editor-popup';
import { TextSelection } from 'prosemirror-state';
import { store } from '../../store';
import { Provider } from 'react-redux';

export class FigureCitationNodeView implements NodeView {
  dom?: HTMLAnchorElement;
  contentDOM?: HTMLSpanElement;
  figureEditorContainer: HTMLDivElement;

  constructor(private node: ProsemirrorNode, private view: EditorView, private getPos: () => number) {
    this.dom = document.createElement('a');
    this.dom.style.cursor = 'pointer';
    this.dom.addEventListener('click', this.selectNode);
    this.contentDOM = document.createElement('span');
    this.dom.appendChild(this.contentDOM);
    if (this.node.attrs.figIds === null) {
      this.open();
    }
  }

  selectNode = (): void => {
    this.dom.classList.add('ProseMirror-selectednode');
    this.open();
  };

  deselectNode(): void {
    this.dom.classList.remove('ProseMirror-selectednode');
    this.close();
  }

  update(node: ProsemirrorNode): boolean {
    this.node = node;
    return true;
  }

  open(): void {
    this.figureEditorContainer = this.view.dom.parentNode.appendChild(document.createElement('div'));

    ReactDOM.render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <FigureCitationEditorPopup
            selectedIds={this.node.attrs.figIds || []}
            anchorEl={this.dom}
            onClose={this.close}
            onChange={this.handleChange}
          />
        </ThemeProvider>
      </Provider>,
      this.figureEditorContainer
    );
  }

  stopEvent(): boolean {
    return true;
  }

  ignoreMutation(): boolean {
    return true;
  }

  close = (): void => {
    this.dom.classList.remove('ProseMirror-selectednode');
    if (this.figureEditorContainer) {
      ReactDOM.unmountComponentAtNode(this.figureEditorContainer);
      this.figureEditorContainer.parentNode.removeChild(this.figureEditorContainer);
      this.figureEditorContainer = null;
    }
  };

  handleChange = (figures: FiguresListEntry[]): void => {
    const ids = figures.map(({ id }) => id);
    const change = this.view.state.tr.setNodeMarkup(this.getPos(), null, { figIds: ids });

    if (this.node.textContent === '???' && figures.length > 0) {
      const nodeContent = (figures[0].name || UNLABELLED_FIGURE_TEXT).replace(/\.$/, '');
      change
        .setSelection(TextSelection.create(change.doc, this.getPos() + 1, this.getPos() + this.node.nodeSize - 1))
        .replaceSelectionWith(this.view.state.schema.text(nodeContent));
    } else if (figures.length === 0) {
      change
        .setSelection(TextSelection.create(change.doc, this.getPos() + 1, this.getPos() + this.node.nodeSize - 1))
        .replaceSelectionWith(this.view.state.schema.text('???'));
    }

    this.view.dispatch(change);
  };
}
