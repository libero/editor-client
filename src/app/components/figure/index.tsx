import React from 'react';
import { NodeView, EditorView } from 'prosemirror-view';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import ReactDOM from 'react-dom';
import { ThemeProvider } from '@material-ui/core/styles';
import { Transaction, TextSelection, NodeSelection } from 'prosemirror-state';

import { StepMap } from 'prosemirror-transform';

import { theme } from 'app/styles/theme';
import { FigureEditor, FigureEditorHandle } from 'app/components/figure/figure-editor';

export class FigureNodeView implements NodeView {
  dom?: HTMLElement;
  figureEditor: React.RefObject<FigureEditorHandle>;

  constructor(
    private node: ProsemirrorNode,
    private view: EditorView,
    private getPos: () => number,
    private isContainerActive: () => boolean
  ) {
    this.dom = document.createElement('section');
    this.figureEditor = React.createRef();
    ReactDOM.render(
      <ThemeProvider theme={theme}>
        <FigureEditor
          ref={this.figureEditor}
          figureNode={this.node}
          onDelete={this.handleDelete}
          onLabelChange={this.handleLabelChange}
          onImageChange={this.handleImageChange}
          onNodeChange={this.handleContentChange}
          onSelectionChange={this.handleSelectionChange}
        />
      </ThemeProvider>,
      this.dom
    );
  }

  handleLabelChange = (label: string) => {
    const newAttributes = { label, img: this.node.attrs.img };
    const change = this.view.state.tr.setNodeMarkup(this.getPos(), null, newAttributes);
    this.view.dispatch(change);
  };

  handleImageChange = (img: string) => {
    const newAttributes = { img, label: this.node.attrs.label };
    const change = this.view.state.tr.setNodeMarkup(this.getPos(), null, newAttributes);
    this.view.dispatch(change);
  };

  handleContentChange = (nodeViewChange: Transaction, offset: number) => {
    this.view.dispatch(this.transformNodeViewChanges(nodeViewChange, offset));
  };

  handleDelete = () => {
    const change = this.view.state.tr
      .setSelection(new NodeSelection(this.view.state.doc.resolve(this.getPos())))
      .deleteSelection();
    this.view.dispatch(change);
  };

  handleSelectionChange = (anchor: number, head: number) => {
    const offset = this.getPos() + 1;
    const start = Math.min(anchor, head);
    const end = Math.max(anchor, head);
    const selection = TextSelection.create(this.view.state.doc, start + offset, end + offset);
    if (!selection.eq(this.view.state.selection)) {
      const selectionChange = this.view.state.tr.setSelection(selection);
      this.view.dispatch(selectionChange);
      this.view.dom.dispatchEvent(new FocusEvent('focus'));
    }
  };

  update(node: ProsemirrorNode) {
    if (!node.sameMarkup(this.node)) {
      this.node = node;
    }
    this.figureEditor.current.updateContent(node);
    if (this.isContainerActive() && !this.figureEditor.current.hasFocus()) {
      this.figureEditor.current.focusFromSelection(this.view.state.selection, this.getPos());
    }
    return true;
  }

  stopEvent(evt) {
    return this.dom.contains(evt.target);
  }

  ignoreMutation() {
    return true;
  }

  private transformNodeViewChanges(change: Transaction, additionalOffset: number): Transaction {
    const parentChange = this.view.state.tr;
    const offsetMap = StepMap.offset(this.getPos() + 2 + additionalOffset);
    const steps = change.steps;
    for (let i = 0; i < steps.length; i++) {
      parentChange.step(steps[i].map(offsetMap));
    }
    return parentChange;
  }
}
