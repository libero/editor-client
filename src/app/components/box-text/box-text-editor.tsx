import React, { useCallback, useRef, useImperativeHandle } from 'react';
import { EditorState, Transaction, TextSelection } from 'prosemirror-state';
import DeleteIcon from '@material-ui/icons/Delete';
import { IconButton } from '@material-ui/core';

import { RichTextEditor } from 'app/components/rich-text-editor';
import { useBoxTextEditorStyles } from 'app/components/box-text/styles';

export interface BoxTextEditorHandle {
  updateContent(editorState: EditorState): void;
  focus(): void;
}

interface BoxTextEditorProps {
  editorState: EditorState;
  onDelete(): void;
  onNodeChange(change: Transaction): void;
  onSelectionChange(from: number, anchor: number): void;
}

export const BoxTextEditor = React.forwardRef((props: BoxTextEditorProps, ref) => {
  const { editorState, onNodeChange, onSelectionChange, onDelete } = props;
  const classes = useBoxTextEditorStyles();
  const boxTextRef = useRef(null);

  useImperativeHandle(ref, () => ({
    updateContent: (editorState: EditorState) => {
      const { $head, $anchor } = boxTextRef.current.editorView.state.selection;
      const selection = TextSelection.create(editorState.doc, $anchor.pos, $head.pos);
      boxTextRef.current.updateEditorState(editorState.apply(editorState.tr.setSelection(selection)));
    },
    focus: () => {
      boxTextRef.current.focus();
    }
  }));

  const handleContentChange = useCallback(
    (change: Transaction) => {
      if (change.docChanged) {
        onNodeChange(change);
      } else {
        onSelectionChange(change.selection.$anchor.pos, change.selection.$head.pos);
      }
    },
    [onNodeChange, onSelectionChange]
  );

  return (
    <div className={classes.boxContainer}>
      <div className="box-text">
        <RichTextEditor
          ref={boxTextRef}
          isActive={false}
          variant="outlined"
          label="Box text"
          editorState={editorState}
          onChange={handleContentChange}
        ></RichTextEditor>
      </div>
      <IconButton classes={{ root: classes.editButton }} onClick={onDelete}>
        <DeleteIcon fontSize="small" />
      </IconButton>
    </div>
  );
});
