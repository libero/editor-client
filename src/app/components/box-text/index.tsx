import React from 'react';
import { NodeView, EditorView } from 'prosemirror-view';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import ReactDOM from 'react-dom';
import { ThemeProvider } from '@material-ui/core/styles';
import { NodeSelection } from 'prosemirror-state';

import { BoxTextEditor, BoxTextEditorHandle } from './box-text-editor';
import { theme } from '../../styles/theme';
import { NodeViewContext } from '../../utils/view.utils';

export class BoxTextNodeView implements NodeView {
  dom?: HTMLElement;
  boxTextEditor: React.RefObject<BoxTextEditorHandle>;

  constructor(private node: ProsemirrorNode, private view: EditorView, private getPos: () => number) {
    this.dom = document.createElement('section');
    this.boxTextEditor = React.createRef();
    ReactDOM.render(
      <ThemeProvider theme={theme}>
        <NodeViewContext.Provider
          value={{
            view: this.view,
            getPos: this.getPos,
            getNode: () => this.node
          }}
        >
          <BoxTextEditor ref={this.boxTextEditor} node={this.node} onDelete={this.handleDelete} />
        </NodeViewContext.Provider>
      </ThemeProvider>,
      this.dom
    );

    this.dom.addEventListener('drop', this.handleDrop, true);
  }

  handleDrop = (event: Event): void => {
    event.stopPropagation();
    event.preventDefault();
  };

  handleDelete = (): void => {
    const change = this.view.state.tr
      .setSelection(new NodeSelection(this.view.state.doc.resolve(this.getPos())))
      .deleteSelection();
    this.view.dispatch(change);
  };

  update(node: ProsemirrorNode): boolean {
    this.node = node;
    this.boxTextEditor.current.updateContent(node);
    return true;
  }

  stopEvent(evt): boolean {
    return this.dom.contains(evt.target) || /drag/.test(evt.type);
  }

  destroy(): void {
    this.dom.removeEventListener('drop', this.handleDrop, true);
  }

  ignoreMutation(): boolean {
    return true;
  }
}
