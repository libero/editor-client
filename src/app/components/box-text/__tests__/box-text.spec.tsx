import React from 'react';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import { ReplaceStep } from 'prosemirror-transform';

import { createBodyState } from 'app/models/body';
import { EditorView } from 'prosemirror-view';
import { BoxTextNodeView } from 'app/components/box-text';
import { BoxTextEditor } from 'app/components/box-text/box-text-editor';

jest.mock('prosemirror-view');
jest.mock('app/components/box-text/box-text-editor', () => ({
  BoxTextEditor: class BoxTextEditorClass extends jest.requireActual('react').Component {
    static updateContentSpy = jest.fn();
    updateContent(...args) {
      BoxTextEditorClass.updateContentSpy(...args);
    }
    render() {
      return <div data-cmp="BoxTextEditor" />;
    }
  }
}));

jest.mock('@material-ui/core/styles', () => {
  return {
    ThemeProvider: ({ children }) => <>{children}</>,
    createMuiTheme: jest.fn()
  };
});

describe('BoxText node view', () => {
  it('should render figure node view', () => {
    const figureNode = givenBoxTextNode();
    const view = new EditorView(null, {
      state: null,
      dispatchTransaction: jest.fn()
    });

    view.state = createBodyState(document.createElement('body'), '');
    const nodeView = new BoxTextNodeView(figureNode, view, jest.fn().mockReturnValue(0));

    expect(nodeView.dom.innerHTML).toMatchSnapshot();
  });

  it('should delete node', () => {
    const figureNode = givenBoxTextNode();
    const view = new EditorView(null, {
      state: null,
      dispatchTransaction: jest.fn()
    });

    view.state = createBodyState(document.createElement('body'), '');
    const nodeView = new BoxTextNodeView(figureNode, view, jest.fn().mockReturnValue(0));
    jest.spyOn(view, 'dispatch');
    nodeView.handleDelete();
    expect(view.dispatch['mock'].calls[0][0].steps[0]).toBeInstanceOf(ReplaceStep);
  });

  it('should update node', () => {
    const figureNode = givenBoxTextNode();
    const view = new EditorView(null, {
      state: null,
      dispatchTransaction: jest.fn()
    });

    view.state = createBodyState(document.createElement('body'), '');
    const nodeView = new BoxTextNodeView(figureNode, view, jest.fn().mockReturnValue(0));
    const newNode = givenBoxTextNode();
    nodeView.update(newNode);
    expect(BoxTextEditor['updateContentSpy']).toBeCalledWith(newNode);
  });

  it('should stop event if target is part of node view', () => {
    const figureNode = givenBoxTextNode();
    const view = new EditorView(null, {
      state: null,
      dispatchTransaction: jest.fn()
    });

    view.state = createBodyState(document.createElement('body'), '');
    const nodeView = new BoxTextNodeView(figureNode, view, jest.fn().mockReturnValue(0));
    expect(nodeView.stopEvent(new Event('click'))).toBeFalsy();
    const evt = new Event('click');
    nodeView.dom.firstChild.dispatchEvent(evt);
    expect(nodeView.stopEvent(evt)).toBeTruthy();
  });

  it('should ignore mutations', () => {
    const figureNode = givenBoxTextNode();
    const view = new EditorView(null, {
      state: null,
      dispatchTransaction: jest.fn()
    });

    view.state = createBodyState(document.createElement('body'), '');
    const nodeView = new BoxTextNodeView(figureNode, view, jest.fn().mockReturnValue(0));
    expect(nodeView.ignoreMutation()).toBeTruthy();
  });
});

function givenBoxTextNode(): ProsemirrorNode {
  const bodyEditorState = createBodyState(document.createElement('div'), '');
  return bodyEditorState.schema.nodes.boxText.createAndFill();
}
