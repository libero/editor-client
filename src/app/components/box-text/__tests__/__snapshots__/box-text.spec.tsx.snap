// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoxText node view should render figure node view 1`] = `"<div data-cmp=\\"BoxTextEditor\\"></div>"`;
